### PR TITLE
Improve documentation for mocks

### PR DESCRIPTION
**Pull request by elenaoat**
Add more meaningful examples and correct a few grammar mistakes.

### CODE REVIEW

File: Doc/library/unittest.mock-examples.rst [hunk 1]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
...
-    ...     async def __aexit__(self, exc_type, exc, tb):
-    ...         pass
-    ...
-    >>> mock_instance = MagicMock(AsyncContextManager())  # AsyncMock also works here
     >>> async def main():
-    ...     async with mock_instance as result:
-    ...         pass
-    ...
+    ...     async with MagicMock() as async_context_manager: # Replacing MagicMock with AsyncMock would also work
+    ...         print('Inside async context manager')
```
Comment (lisroach) [inline Doc/library/unittest.mock-examples.rst:323]: I think you need a newline with `...` here

File: Doc/library/unittest.mock-examples.rst [hunk 2]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
...
+    >>> asyncio.run(main())
+    Inside async context manager
+
+This would not work with Mock class, though, because it doesn't have the pre-assigned ``__aexit__`` and ``__aenter__`` methods.
+
+    >>> async def main():
+    ...     async with Mock() as context_manager:
+    ...         print('Inside async context manager')
```
Comment (lisroach) [inline Doc/library/unittest.mock-examples.rst:333]: Same as above, need `...` here

File: Doc/library/unittest.mock-examples.rst [hunk 3]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
 Since Python 3.8, ``AsyncMock`` and ``MagicMock`` have support to mock
 :ref:`async-context-managers` through ``__aenter__`` and ``__aexit__``.
 By default, ``__aenter__`` and ``__aexit__`` are ``AsyncMock`` instances that
-return an async function.
+return an async function. So you can do:
 
-    >>> class AsyncContextManager:
```
Comment (lisroach) [inline Doc/library/unittest.mock-examples.rst:305]: Is this documentation made more clear by removing this example? The goal here was to show how someone could mock an asynchronous context manager that they wrote using MagicMock (or AsyncMock). You've changed it to show MagicMock can act as an asynchronous context manager, which I think without the context of mocking a real object doesn't demonstrate a useful scenario.

File: Doc/library/unittest.mock-examples.rst [hunk 4]
```diff
@@ -208,15 +208,13 @@ If you need an attribute setting on your mock, just do it:
...
-return a list, then we have to configure the result of the nested call.
+``mock.connection.cursor().execute("SELECT 1")``.
 
 We can use :data:`call` to construct the set of calls in a "chained call" like
 this for easy assertion afterwards:
 
     >>> mock = Mock()
-    >>> cursor = mock.connection.cursor.return_value
-    >>> cursor.execute.return_value = ['foo']
+    >>> mock.connection.cursor.return_value.execute.return_value = ['foo']
```
Comment (lisroach) [inline Doc/library/unittest.mock-examples.rst]: I'm not sure I think this is more clear, I do write my actual code this way but is it harder to read this way? Some new users struggle to reason about a line like this.

**Review by lisroach — CHANGES_REQUESTED** [review]

File: Doc/library/unittest.mock-examples.rst [hunk 5]
```diff
@@ -208,15 +208,13 @@ If you need an attribute setting on your mock, just do it:
     3
 
 Sometimes you want to mock up a more complex situation, like for example
-``mock.connection.cursor().execute("SELECT 1")``. If we wanted this call to
-return a list, then we have to configure the result of the nested call.
```
Comment (ezio-melotti) [inline Doc/library/unittest.mock-examples.rst:212]: I would rephrase this as `If we wanted this call to return a list, then we have to specify the return value of the two calls.`.
I think here "nested" is used incorrectly, because even though the mocks might be considered nested, the calls in the expressions are not.  If we want to make this more explicit, the two calls (cursor() and execute()) could be mentioned explicitly.

File: Doc/library/unittest.mock-examples.rst (hunk 4 above)
Comment (ezio-melotti) [reply Doc/library/unittest.mock-examples.rst]: I find this whole example somewhat confusing, since ISTM it's trying to explain two things at once:
1) how to mock a complex expression to return a certain value;
2) how to use `call` and `call_list()`,
I'm also not sure what the point of using `call` is, just to prove that both methods got called? If that's the case, is not very clear, and it does more harm than good IMHO.

The second line (`cursor = mock.connection.cursor.return_value`) is a bit unusual since we usually set the `return_value`, whereas here we are just getting a reference to the already existing default `return_value` (a mock).  This could be clarified.  I would just leave the shorthand suggested by Elena, and explain how it works.

To summarize, the whole thing could look like:
```rest
Sometimes you want to mock up a more complex expression, such as
``db.connection.cursor().execute("SELECT 1")``.  If we wanted this
call to return ``['foo']``, then we can do:

   >>> db = Mock()
   >>> db.connection.cursor.return_value.execute.return_value = ['foo']

Since we didn't specify a ``return_value`` for ``mock.connection.cursor()``,
a new mock will be created by default.  This new mock will have an
``execute`` method that returns ``['foo']`` when called::

   >>> cursor = db.connection.cursor()
   >>> cursor  # check that cursor is a new mock with an execute method
   <Mock name='db.connection.cursor()' id='140217681027816'>
   >>> cursor.execute("SELECT 1")
   ['foo']

This is equivalent to executing the initial expression::

   >>> db.connection.cursor().execute("SELECT 1")
   ['foo']
```

I would remove all the part about `call` -- if we want to explain how it works, we could do it in a separate entry.  I don't think in a real-world situation I would use it to check that both methods are called -- I would probably just check the final result and perhaps `db.connection.cursor.assert_called()`.  For this case, I don't think including these checks adds more value.

File: Doc/library/unittest.mock-examples.rst (hunk 3 above)
Comment (ezio-melotti) [reply Doc/library/unittest.mock-examples.rst:305]: The confusing part is that the CM class is defined together with the code that mocks it.  The user might also want to mock a CM that they didn't define, but obtain through a call (e.g. to `open`).

In the first case, it might be enough to imply the existence of the CM by doing `from mymodule import AsyncContextManager`, and describing how `MagicMock(AsyncContextManager())` automatically find and mocks `__aenter__` and `__aexit__` with `AsyncMock`s.

In the second case, what Elena did is OK unless we want a more realistic example.

File: Doc/library/unittest.mock-examples.rst [hunk 6]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
...
-    ...         return self
-    ...     async def __aexit__(self, exc_type, exc, tb):
-    ...         pass
-    ...
-    >>> mock_instance = MagicMock(AsyncContextManager())  # AsyncMock also works here
     >>> async def main():
-    ...     async with mock_instance as result:
-    ...         pass
-    ...
+    ...     async with MagicMock() as async_context_manager: # Replacing MagicMock with AsyncMock would also work
```
Comment (ezio-melotti) [inline Doc/library/unittest.mock-examples.rst]: I liked more the shorter original comment.

File: Doc/library/unittest.mock-examples.rst [hunk 7]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
...
+    Inside async context manager
+
+This would not work with Mock class, though, because it doesn't have the pre-assigned ``__aexit__`` and ``__aenter__`` methods.
+
+    >>> async def main():
+    ...     async with Mock() as context_manager:
+    ...         print('Inside async context manager')
     >>> asyncio.run(main())
-    >>> mock_instance.__aenter__.assert_awaited_once()
-    >>> mock_instance.__aexit__.assert_awaited_once()
```
Comment (ezio-melotti) [inline Doc/library/unittest.mock-examples.rst:318]: I think these 2 lines should be kept.

File: Doc/library/unittest.mock-examples.rst [hunk 8]
```diff
@@ -300,28 +298,29 @@ Mocking asynchronous context manager
...
     >>> async def main():
-    ...     async with mock_instance as result:
-    ...         pass
-    ...
+    ...     async with MagicMock() as async_context_manager: # Replacing MagicMock with AsyncMock would also work
+    ...         print('Inside async context manager')
+    >>> asyncio.run(main())
+    Inside async context manager
+
+This would not work with Mock class, though, because it doesn't have the pre-assigned ``__aexit__`` and ``__aenter__`` methods.
```
Comment (ezio-melotti) [inline Doc/library/unittest.mock-examples.rst]: I'm not sure this needs an example, the sentence is enough.
(Also the line is too long.)

**Review by ezio-melotti — CHANGES_REQUESTED** [review]

File: Doc/library/unittest.mock-examples.rst (hunk 3 above)
Comment (elenaoat) [reply Doc/library/unittest.mock-examples.rst:305]: In my opinion defining the `AsyncContextManager` class here is kind of useless. All of its methods will be mocked by MagicMock, so we might as well forget of its existence, or am I missing something here? What I mean is that `mock_instance = MagicMock(AsyncContextManager())` is equivalent to `mock_instance = MagicMock()`

File: Doc/library/unittest.mock-examples.rst (hunk 7 above)
Comment (elenaoat) [reply Doc/library/unittest.mock-examples.rst:318]: The Mock doesn't have these two methods, that's why it will fail.

File: Doc/library/unittest.mock-examples.rst (hunk 3 above)
Comment (lisroach) [reply Doc/library/unittest.mock-examples.rst:305]: @elenaoat I don't agree that because the custom context manager is mocked it is not relevant. Mock is used to mock objects that exist in code to have different behavior for the sake of testing. I cannot come up with a use case for a mock object as an asynchronous context manager that is not specced with any real piece of code.  What problem is it solving?
Maybe the test could be more verbose by adding a return value or side effect on the mock object, but I don't think it makes sense to ever use it without a spec.

 And `mock_instance = MagicMock(AsyncContextManager())` is not the same as `mock_instance =MagicMock()` off the bat:

```
>>> dir(MagicMock())
['assert_any_call', 'assert_called', 'assert_called_once', 'assert_called_once_with', 'assert_called_with', 'assert_has_calls', 'assert_not_called', 'attach_mock', 'call_args', 'call_args_list', 'call_count', 'called', 'configure_mock', 'method_calls', 'mock_add_spec', 'mock_calls', 'reset_mock', 'return_value', 'side_effect']
>>> dir(MagicMock(AsyncContextManager))
['__aenter__', '__aexit__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', 'assert_any_call', 'assert_called', 'assert_called_once', 'assert_called_once_with', 'assert_called_with', 'assert_has_calls', 'assert_not_called', 'attach_mock', 'call_args', 'call_args_list', 'call_count', 'called', 'configure_mock', 'method_calls', 'mock_add_spec', 'mock_calls', 'reset_mock', 'return_value', 'side_effect']
```

I agree that it is weird to define them together in the same file, but I think that might be a limitation of doctest. I assumed we would not be able to import `AsyncContextManager` without it failing, because `mymodule` doesn't actually exist? I might be wrong about how this works though. If we can import it I think that is a good idea, maybe make the name more explicit like `MyCustomAsyncContextManager`. 

If we can't import it I still think it is better to build it in here then exclude it all together, maybe we can add a comment block around it, implying it is in a different file without actually being in a different file?

**Comment by lisroach** [comment]
Thanks for the ping @csabella this got buried for me. I'd like to hear @tirkarthi opinions on this too, since he was a significant author for these docs.

File: Doc/library/unittest.mock-examples.rst [hunk 9]
```diff
@@ -207,26 +207,29 @@ If you need an attribute setting on your mock, just do it:
...
-return a list, then we have to configure the result of the nested call.
+Sometimes you want to mock up a more complex expression, such as
+``db.connection.cursor().execute("SELECT 1")``. If we wanted this
+call to return a list, e.g. ``['foo']``, we'd have to specify the return value
+for both of the calls:
 
-We can use :data:`call` to construct the set of calls in a "chained call" like
-this for easy assertion afterwards:
+   >>> db = Mock()
+   >>> db.connection.cursor.return_value.execute.return_value = ['foo']
```
Comment (tirkarthi) [inline Doc/library/unittest.mock-examples.rst:216]: Though the earlier example was slightly verbose I think it's clear on setting nested mock return values. Personally I don't prefer much nested chains in the docs so I would suggest keeping the old example and then adding a note with this line that nested mocks return value could be set with this one liner as alternative.

File: Doc/library/unittest.mock.rst [hunk 10]
```diff
@@ -527,10 +543,19 @@ the *new_callable* argument to :func:`patch`.
...
+        You can call the return value just like any other mock:
+
             >>> mock.return_value()
             <Mock name='mock()()' id='...'>
-            >>> mock.return_value.assert_called_with()
+
+        You can set attributes on the return value too:
+
+            >>> mock.return_value.attribute = 'some random attribute value'
+            >>> mock.return_value.attribute
```
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst:558]: Instead of demonstrating the attribute to be present on the return_value attribute itself we the below reads more useful showing the attribute to be set on the return value itself as call is made.

```suggestion
            >>> mock().attribute
```

File: Doc/library/unittest.mock.rst [hunk 11]
```diff
@@ -148,10 +148,26 @@ easiest way of using magic methods is with the :class:`MagicMock` class. It
 allows you to do things like:
 
     >>> mock = MagicMock()
-    >>> mock.__str__.return_value = 'foobarbaz'
-    >>> str(mock)
-    'foobarbaz'
-    >>> mock.__str__.assert_called_with()
+    >>> len(mock)
```
Comment (tirkarthi) [inline Doc/library/unittest.mock.rst:151]: I would suggest keeping the old example setting `return_value` on `__str__` though there is a different example below setting `__str__.return_value` with a mock. May I know the purpose of changing it to `len`? Is it show the default value for some magic methods?

**Review by tirkarthi — COMMENTED** [review]
Apologies, I thought I submitted the review. I didn't know it was pending.

**Comment by tirkarthi** [comment]
See also https://bugs.python.org/issue40406 with reports about the nested mocking workflow to be confusing,

**Comment by erlend-aasland** [comment]
@elenaoat: are you planning on following up this PR?

**Comment by willingc** [comment]
I'm going to close this PR as it has been pending over six months. 

Thanks @elenaoat and reviewers for your effort.
